client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -246,6 +259,18 @@ public SendShuffleDataResult sendShuffleData(String appId, List<ShuffleBlockInfo
     // which is minimum number when there is at most *replicaWrite - replica* sending server failures.
     for (ShuffleBlockInfo sbi : shuffleBlockInfoList) {
       List<ShuffleServerInfo> allServers = sbi.getShuffleServerInfos();
+      if (replica > 1 && !shuffleServerBlacklist.isEmpty()) {
+        // allServers is shared and is not thread safe
+        synchronized (this) {
+          for (int i = allServers.size() - 2; i >= 0; i--) {
+            ShuffleServerInfo serverInfo = allServers.get(i);
+            if (shuffleServerBlacklist.contains(serverInfo)) {
```

Comment:
What will it happen if all the servers are added the blacklist?

Reply:
If all the servers are added the blacklist, the sending order will be third -> second -> first, it all servers are unavailable, the task will be failed. And if some servers become available, it will be removed from the blacklist.